pkg/sql/colexec/agg/variance/variance.go

```diff
@@ -0,0 +1,221 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package Variance
+
+import (
+	"math"
+	"reflect"
+	"strconv"
+	"unsafe"
+
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+)
+
+func ReturnType(_ []types.Type) types.Type {
+	return types.New(types.T_float64, 0, 0, 0)
+}
+
+func String(b []byte) (s string) {
+	pbytes := (*reflect.SliceHeader)(unsafe.Pointer(&b))
+	pstring := (*reflect.StringHeader)(unsafe.Pointer(&s))
+	pstring.Data = pbytes.Data
+	pstring.Len = pbytes.Len
+	return
+}
+
+/*
+	case types.Decimal64:
+		res, _ := strconv.ParseFloat(String(v.Decimal64ToString(variance.inputType.Scale)), 64)
+		return res
+	case types.Decimal128:
+		res, _ := strconv.ParseFloat(String(v.Decimal128ToString(variance.inputType.Scale)), 64)
+*/
+
+//New1 is used to Created a Variance which supports float,int,uint
+func New1[T1 types.Floats | types.Ints | types.UInts]() *Variance[T1] {
+	return &Variance[T1]{}
+}
+
+//New2 is used to Created a Variance which supports decimal64
+//if is decimal,you need to give pricision
+func New2() *Variance2 {
+	return &Variance2{inputType: types.New(types.T_decimal64, 0, 0, 0)}
+}
+
+//New2 is used to Created a Variance which supports decimal64
+//if is decimal,you need to give pricision
+func New3() *Variance3 {
+	return &Variance3{inputType: types.New(types.T_decimal64, 0, 0, 0)}
+}
+
+func (variance *Variance[T1]) Grows(sizes int) {
+	if len(variance.avgs) == 0 {
+		variance.avgs = make([]float64, sizes)
+		variance.sizes = make([]float64, sizes)
+	} else {
+		for i := 0; i < sizes; i++ {
+			var a float64
+			variance.avgs = append(variance.avgs, a)
+			variance.sizes = append(variance.sizes, 0)
+		}
+	}
+}
+
+func (variance *Variance[T1]) Eval(vs []float64) []float64 {
+	for i, v := range vs {
+		avg := (variance.avgs[i]) / (variance.sizes[i])
+		vs[i] = (v)/(variance.sizes[i]) - math.Pow(float64(avg), 2)
+	}
+	return vs
+}
+
+func (variance *Variance[T1]) Merge(groupIndex1, groupIndex2 int64, x, y float64, IsEmpty1 bool, IsEmpty2 bool, agg any) (float64, bool) {
+	variance2 := agg.(*Variance[T1])
+	if IsEmpty1 && !IsEmpty2 {
+		variance.avgs[groupIndex1] = variance2.avgs[groupIndex2]
+		variance.sizes[groupIndex1] = variance.sizes[groupIndex2]
+		return y, false
+	} else if IsEmpty2 && !IsEmpty1 {
+		return x, false
+	} else if IsEmpty1 && IsEmpty2 {
+		return x, true
+	} else {
+		variance.sizes[groupIndex1] += variance2.sizes[groupIndex2]
+		variance.avgs[groupIndex1] += variance2.avgs[groupIndex2]
+		return (x + y), false
+	}
+}
+
+func (variance *Variance[T1]) Fill(groupIndex int64, v1 T1, v2 float64, z int64, IsEmpty bool, hasNull bool) (float64, bool) {
+	if hasNull {
+		return v2, IsEmpty
+	} else if IsEmpty {
+		f1 := float64(v1)
+		variance.avgs[groupIndex] = any(f1 * float64(z)).(float64)
+		variance.sizes[groupIndex] += float64(z)
+		return math.Pow(f1, 2) * float64(z), false
+	}
+	f1 := float64(v1)
+	f2 := float64(v2)
```

Comment:
I am not sure, but your variable naming totally confused me.   sizes seems to be count.    avgs is definitely not avgs, it seems to be sum(x).    Where is sum(x^2)?    You returned it?

Reply:
sum(x^2） is returned by me to stored in the slice vs. sizes is count, avgs is sum(x) (I used avgs because it';ll used to caculate the avgs at last), I think I need to modify the names.